### PR TITLE
Avoid sharing custom serializers among runtime instances

### PR DESCRIPTION
**Pull request by Lujie1996**
Different corfu runtime instances can have different custom serializers based on
the opened tables. The existing serializers is a static member shared by all
corfu runtimes in a JVM which is problematic when multiple runtimes register
their own custom serialzier.

## Overview

Description:

Why should this be merged: 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

**Review by xnull — APPROVED** [review]

**Review by PavelZaytsev — APPROVED** [review]

File: infrastructure/src/test/java/org/corfudb/infrastructure/management/FailureDetectorTest.java [hunk 1]
```diff
@@ -40,7 +40,7 @@ public void testPollRound() {
         );
         Duration time = Duration.ofMillis(System.currentTimeMillis() - start);
         assertThat(time).isGreaterThan(Duration.ofMillis(450));
-        assertThat(time).isLessThan(Duration.ofSeconds(2));
+        assertThat(time).isLessThan(Duration.ofSeconds(3));
```
Comment (Maithem) [inline infrastructure/src/test/java/org/corfudb/infrastructure/management/FailureDetectorTest.java:43]: Can you do this in a different PR ?

**Review by Maithem — APPROVED** [review]

File: infrastructure/src/test/java/org/corfudb/infrastructure/management/FailureDetectorTest.java (hunk 1 above)
Comment (Lujie1996) [reply infrastructure/src/test/java/org/corfudb/infrastructure/management/FailureDetectorTest.java]: Done

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java [hunk 2]
```diff
@@ -85,6 +85,7 @@ public Object getPayload(CorfuRuntime runtime) {
         if (value instanceof LogEntry) {
             if (!Address.isAddress(((LogEntry) value).getGlobalAddress())) {
                 ((LogEntry) value).setGlobalAddress(getGlobalAddress());
+                ((LogEntry) value).setRuntime(runtime);
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java:88]: Can you just pass the serializer here directly?

**Review by Maithem — COMMENTED** [review]
Instead of passing the CorfuRuntime to every method that needs to serialize/deserialize I suggest we just pass the ISerializer object directly.

**Review by achinmishra — APPROVED** [review]
lgtm

File: runtime/src/main/java/org/corfudb/util/serializer/Serializers.java [hunk 3]
```diff
@@ -41,14 +41,14 @@ public static final ISerializer getDefaultSerializer() {
         serializersMap.put(QUEUE_SERIALIZER.getType(), QUEUE_SERIALIZER);
     }
 
-    private static final Map<Byte, ISerializer> customSerializers = new HashMap<>();
+    private final Map<Byte, ISerializer> customSerializers = new HashMap<>();
```
Comment (pankti-m) [inline runtime/src/main/java/org/corfudb/util/serializer/Serializers.java:44]: should we still keep it as a map?  can a runtime use >1 custom serializers?

**Review by pankti-m — APPROVED** [review]

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java (hunk 2 above)
Comment (Lujie1996) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java:88]: I thought about it, but then I found that SMREntry calls `serializerType.deserialize(objBuf, runtime)` and JsonSerializer will need runtime to build an object

`rt.getObjectsView().build()
                        .setStreamID(new UUID(b.readLong(), b.readLong()))
                        .setType((Class<? extends ICorfuSMR>) Class.forName(smrClassName))
                        .open();`

File: runtime/src/main/java/org/corfudb/util/serializer/Serializers.java (hunk 3 above)
Comment (Lujie1996) [reply runtime/src/main/java/org/corfudb/util/serializer/Serializers.java:44]: I think it's a map now, right?

Comment (pankti-m) [reply runtime/src/main/java/org/corfudb/util/serializer/Serializers.java:44]: I meant, can we remove the map?  Why should it be a map?
